modules/core/04-channel/keeper/upgrade.go

```diff
@@ -326,3 +325,73 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
 		LatestSequenceSend: nextSequenceSend - 1,
 	}, nil
 }
+
+// abortUpgrade will restore the channel state and flush status to their pre-upgrade state so that upgrade is aborted.
+// any unnecessary state is deleted. An error receipt is written, and the OnChanUpgradeRestore callback is called.
+func (k Keeper) abortUpgrade(ctx sdk.Context, portID, channelID string, err error) error {
+	if err == nil {
+		return errorsmod.Wrap(types.ErrInvalidUpgradeError, "cannot abort upgrade handshake with nil error")
+	}
+
+	upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrUpgradeNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if err := k.restoreChannel(ctx, portID, channelID); err != nil {
+		return err
+	}
+
+	// in the case of application callbacks, the error may not be an upgrade error.
+	// in this case we need to construct one in order to write the error receipt.
+	upgradeError, ok := err.(*types.UpgradeError)
+	if !ok {
+		channel, found := k.GetChannel(ctx, portID, channelID)
+		if !found {
+			return errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+		}
+
+		upgradeError = types.NewUpgradeError(channel.UpgradeSequence, err)
+	}
```

Comment:
this feels a bit ugly, open to suggestions on improving this.

Reply:
the channel is already looked up in `restoreChannel`, right? we could move that logic to be scoped within this function such that you don't have to do the lookup twice (would also save gas)

is `restoreChannel()` intended to only be called via `abortUpgrade()`?

Reply:
yep good point, will do a small refactor